content/oci/oci.go

```diff
@@ -192,6 +192,22 @@ func (s *Store) Resolve(ctx context.Context, reference string) (ocispec.Descript
 	return desc, nil
 }
 
+func (s *Store) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	resolvers := s.tagResolver.Map()
+	for reference, desc := range resolvers {
+		if content.Equal(desc, target) {
+			s.tagResolver.Remove(reference)
+		}
+	}
```

Comment:
What if some is pushing and deleting at the same time? There is a race condition.

Reply:
The targetResolver.Map() represent an inconsistent snapshot of the storage content as far as I understand, therefore I'm not sure if a delete of a pushing content would be possible as the tagResolver is actually a syncMap.

Reply:
@carabasdaniel Since it is possible that someone is trying to push and delete at the same time, we have to trade-off some performance for the deletion.

Similar to what I have mentioned in other comments, the current `oci.Store` is append-only and thus it has the benefits that multiple push and fetch can happen at the same time without race conditions.

However, adding the functionality of `Delete()` makes the store mutable, and we have to change the underlying implementation from sync maps to a `RWMutex` to ensure consistency. The drawback is that we no longer can push and delete simultaneously.

Overall, I suggest

1. Drop `Delete()` for `oci.Store`.
2. Develop `oci.MutableStore` with `Delete()` support.

/cc @Wwwsylvia

Reply:
> Overall, I suggest
Drop Delete() for oci.Store.
Develop oci.MutableStore with Delete() support.

It makes sense.

@carabasdaniel I would also like to learn your use cases of deleting content in OCI layout. Is it for GC purpose? Could you share a little bit?

Reply:
Hi @Wwwsylvia, @shizhMSFT,

We have a tool used for building OCI images from rego policies: [policy CLI](https://github.com/opcr-io/policy).

We have been using the previous major version of oras-go to build our containers and interact with upstream OCI compliant registries. In our case we want the user to be able to build and easily remove containers from the local OCI storage, thus the need to have a delete mechanism in place, keeping the same workflow feel as the docker CLI. 

At the moment we use a fork that includes these changes in our v0.2.0 release candidate versions of the policy CLI and it seems to work with our current needs. 

I think having a different store might make sense.

Reply:
Got it. Thank you for sharing this!

Reply:
@Wwwsylvia for additional use case info... I am working on https://github.com/awslabs/soci-snapshotter which uses https://pkg.go.dev/oras.land/oras-go/content to keep content in the local filesystem. Our artifacts are conceptually similar to image and layer metadata; we store one artifact per image we handle, and one artifact per layer in those images, with one-way references (via a list of digests equivalent to image layers) from artifacts of the first type to artifacts of the second type.